### PR TITLE
StaticLLMPipeline: Optimize V-tensors layout

### PR DESCRIPTION
**Pull request by TolyaTalamanov**

### CODE REVIEW

File: samples/cpp/chat_sample/chat_sample.cpp [hunk 1]
```diff
@@ -10,8 +10,10 @@ int main(int argc, char* argv[]) try {
     std::string prompt;
     std::string models_path = argv[1];
 
-    std::string device = "CPU";  // GPU, NPU can be used as well
-    ov::genai::LLMPipeline pipe(models_path, device);
+    std::string device = "NPU";  // GPU, NPU can be used as well
```
Comment (TolyaTalamanov) [inline samples/cpp/chat_sample/chat_sample.cpp:13]: Will be removed

File: src/cpp/src/llm_pipeline_static.cpp [hunk 2]
```diff
@@ -255,7 +459,7 @@ std::optional<NPUDesc> extract_npu_descriptor(ov::Core& core) {
 ov::AnyMap get_baseline_common_config() {
     ov::AnyMap config = {
         { "NPU_COMPILATION_MODE_PARAMS", "compute-layers-with-higher-precision=Sqrt,Power,ReduceMean,Add_RMSNorm" },
-        { "NPUW_DEVICES", "NPU" },
+        { "NPUW_DEVICES", "NPU,CPU" },
```
Comment (TolyaTalamanov) [inline src/cpp/src/llm_pipeline_static.cpp:462]: Remove

File: src/cpp/src/llm_pipeline_static.cpp [hunk 3]
```diff
@@ -449,6 +673,7 @@ void StaticLLMPipeline::setupAndCompileModels(
     drop_cache_dir(prefill_config);
     drop_cache_dir(generate_config);
 
+    ov::save_model(m_kvcache_model, "model-wo-sdpa.xml");
```
Comment (TolyaTalamanov) [inline src/cpp/src/llm_pipeline_static.cpp:676]: Debug

File: src/cpp/src/llm_pipeline_static.cpp [hunk 4]
```diff
@@ -403,38 +636,55 @@ void StaticLLMPipeline::setupAndCompileModels(
...
     const uint32_t kMaxPromptLen = align_to(pop_int_and_cast(properties, "MAX_PROMPT_LEN").value_or(1024u), 64u);
     const uint32_t kMinResponseLen = align_to(pop_int_and_cast(properties, "MIN_RESPONSE_LEN").value_or(128u), 64u);
     KVAxesPosition axes = get_kv_axes(get_model_type_from_json(models_path / "config.json"));
-    m_kvcache_desc = KVCacheDesc { kMaxPromptLen, kMaxPromptLen + kMinResponseLen, 0u, axes.seq_len };
+    m_kvcache_desc = KVCacheDesc { kMaxPromptLen, kMaxPromptLen + kMinResponseLen, 0u, axes.seq_len, false};
     reshape_to_static(m_prefill_model, m_kvcache_desc.max_prompt_size, m_kvcache_desc.max_prompt_size, axes);
     reshape_to_static(m_kvcache_model, 1u, m_kvcache_desc.total_size, axes);
-    // (8) Compile both model
+    // (6) Apply opt layout if applicable
+    const bool disable_opt_layout = pop_or_default<std::string>(properties, "DISABLE_OPT_LAYOUT", "NO") == "YES";
```
Comment (TolyaTalamanov) [inline src/cpp/src/llm_pipeline_static.cpp:668]: So far for debugging purpose, but maybe will be needed as well

File: src/cpp/src/llm_pipeline_static.cpp (hunk 2 above)
Comment (TolyaTalamanov) [reply src/cpp/src/llm_pipeline_static.cpp]: Done

File: src/cpp/src/llm_pipeline_static.cpp (hunk 4 above)
Comment (TolyaTalamanov) [reply src/cpp/src/llm_pipeline_static.cpp:649]: Removed

File: src/cpp/src/llm_pipeline_static.cpp [hunk 5]
```diff
@@ -403,38 +636,55 @@ void StaticLLMPipeline::setupAndCompileModels(
...
+    // NB: Try to apply opt transpose by default for all models that have
+    // KV-cache tensors in format [batch, num_heads, seq_len, emb_size]
+    if (m_kvcache_desc.seq_len == 2 && !disable_opt_layout) {
+        std::cout << "[LOG_DEBUG] Try to apply opt layout" << std::endl;
```
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply src/cpp/src/llm_pipeline_static.cpp]: Removed

File: src/cpp/src/llm_pipeline_static.cpp [hunk 6]
```diff
@@ -20,6 +29,198 @@
...
+            auto matched_node_transpose = node_to_output.at(transpose).get_node_shared_ptr();
+            auto matched_node_matmul    = node_to_output.at(matmul).get_node_shared_ptr();
+
+            auto matched_param     = std::static_pointer_cast<ov::op::v0::Parameter>(matched_node_param);
+            auto matched_concat    = std::static_pointer_cast<ov::op::v0::Concat>(matched_node_concat);
+            auto matched_transpose = std::static_pointer_cast<ov::op::v1::Transpose>(matched_node_transpose);
+            auto matched_matmul    = std::static_pointer_cast<ov::op::v0::MatMul>(matched_node_matmul);
+
+            auto shape = matched_param->get_partial_shape();
+            std::swap(shape[2], shape[3]);
```
Comment (dmatveev) [inline src/cpp/src/llm_pipeline_static.cpp:62]: please comment what are you doing here and add an assert on the shape size.

normally, in the match callback, we have `if()` which checks extra stuff that's not reflected in the pattern structure (like - data types, shapes, layouts, etc.)

Comment (TolyaTalamanov) [reply src/cpp/src/llm_pipeline_static.cpp]: Done

**Comment by dmatveev** [comment]
@TolyaTalamanov put some tickets here idk

File: src/cpp/CMakeLists.txt [hunk 7]
```diff
@@ -56,6 +56,8 @@ ov_genai_build_jinja2cpp()
 
 # Library
 
+find_package(OpenVINO REQUIRED COMPONENTS Runtime Threading)
```
Comment (ilya-lavrenov) [inline src/cpp/CMakeLists.txt:59]: maybe you can change main [find_package](https://github.com/openvinotoolkit/openvino.genai/blob/master/CMakeLists.txt#L52-L60) call in root cmake like it's done in tokenizers?
https://github.com/openvinotoolkit/openvino_tokenizers/blob/5ccd56de5a7da65ad7ea7c5c42cec55be97df312/CMakeLists.txt#L70-L76

File: src/cpp/src/llm_pipeline_static.cpp [hunk 8]
```diff
@@ -20,6 +29,201 @@
...
+            }
+
+            auto new_output_node = decompose(node);
+            ov::replace_node(node, new_output_node);
+            return true;
+        };
+
+        auto m = std::make_shared<ov::pass::pattern::Matcher>(pattern_node, "ScaledDotProductAttentionDecomposition");
+        register_matcher(m, callback);
+ 	}
```
Comment (Wovchena) [inline src/cpp/src/llm_pipeline_static.cpp:116]:
```suggestion
        register_matcher(m, std::move(callback));
    }
```

File: src/cpp/CMakeLists.txt (hunk 7 above)
Comment (TolyaTalamanov) [reply src/cpp/CMakeLists.txt]: Good idea, done

**Review by ilya-lavrenov — APPROVED** [review]

**Review by Wovchena — APPROVED** [review]
Releases branch is frozen, so I don't merge it for now

**Comment by ilya-lavrenov** [comment]
@TolyaTalamanov maybe you can retarget branch to master? and add label `port to LTS`

**Comment by AsyaPronina** [comment]
> @TolyaTalamanov maybe you can retarget branch to master? and add label `port to LTS`

Hello! It is done here: https://github.com/openvinotoolkit/openvino.genai/pull/1232

**Pull request merged**